csharp/source/autowrap/csharp/csharp.d

```diff
@@ -0,0 +1,1061 @@
+    public string namespace;
+    public string functions;
+    public csharpAggregate[string] aggregates;
+
+    public this(string ns) {
+        namespace = ns;
+        functions = string.init;
+    }
+
+    public override string toString() {
+        char[] ret;
+        ret.reserve(1_048_576);
+        foreach(csns; namespaces.byValue()) {
+            ret ~= "namespace " ~ csns.namespace ~ " {" ~newline;
+            ret ~= "    using System;" ~ newline;
+            ret ~= "    using System.Collections.Generic;" ~ newline;
+            ret ~= "    using System.Linq;" ~ newline;
+            ret ~= "    using System.Reflection;" ~ newline;
+            ret ~= "    using System.Runtime.InteropServices;" ~ newline;
+            ret ~= "    using Autowrap;" ~ newline ~ newline;
+            ret ~= "    public static class Functions {" ~ newline;
+            ret ~= csns.functions;
+            ret ~= "    }" ~ newline ~ newline;
+            foreach(agg; aggregates.byValue()) {
+                ret ~= agg.toString();
+            }
+            ret ~= "}" ~ newline;
+        }
+        return cast(immutable)ret;
+    }
+}
+
+private class csharpAggregate {
+    public string name;
+    public bool isStruct;
+    public string constructors;
+    public string functions;
+    public string methods;
+    public string properties;
+
+    public this (string name, bool isStruct) {
+        this.name = name;
+        this.isStruct = isStruct;
+        this.constructors = string.init;
+        this.functions = string.init;
+        this.methods = string.init;
+        this.properties = string.init;
+    }
+
+    public override string toString() {
+        char[] ret;
+        ret.reserve(32_768);
+        if (isStruct) {
+            ret ~= "    [StructLayout(LayoutKind.Sequential)]" ~ newline;
+            ret ~= "    public struct " ~ camelToPascalCase(this.name) ~ " {" ~ newline;
+        } else {
+            ret ~= "    public class " ~ camelToPascalCase(this.name) ~ " : DLangObject {" ~ newline;
+        }
+        if (functions != string.init) ret ~= functions ~ newline;
+        if (constructors != string.init) ret ~= constructors ~ newline;
+        if (methods != string.init) ret ~= methods ~ newline;
+        if (properties != string.init) ret ~= properties;
+        ret ~= "    }" ~ newline ~ newline;
+        return cast(immutable)ret;
+    }
+}
+
+public struct csharpRange {
+    public string constructors = string.init;
+    public string enumerators = string.init;
+    public string functions = string.init;
+    public string getters = string.init;
+    public string setters = string.init; 
+    public string appendValues = string.init;
+    public string appendArrays = string.init;
+    public string sliceEnd = string.init;
+    public string sliceRange = string.init;
+
+    public string toString() {
+        char[] ret;
+        ret.reserve(32_768);
+        ret ~= "    internal class RangeFunctions {" ~ newline;
```

Comment:
Intead of: `ret ~= "..." ~ newline;` it'd be preferable to instead just do `ret ~= "...;` and `.join(newline)` for the final result.

Reply:
.join throws an error when used on a char[] with a string. Also, what is join doing that ~ isn't?

Reply:
Oh, just use `string ret` anyway, there's no advantage to `char[]` unless you'd want to change what's already appended.

> Also, what is join doing that ~ isn't?

Avoiding all the `~ newline` at the end.

Reply:
Actually there is an advantage to char[] because you cannot reserve() strings. So I'm doing one allocation instead of multiple allocations on expansion. For small strings I don't bother, but these output files can be quite lengthy, and reallocations can be much more costly.

And join expects an array of strings in the case which is not something I have to work with. When I tried I got mushroom clouds from the compiler. Also everything in this block is conditional so achieving this will result in more code just to get rid of the `~ newline;` and given that join is just going to append anyways I'm still not seeing the benefit.